// @polymarket/clob-client/dist/rfq-client.js
import { parseUnits } from "viem";
import { COLLATERAL_TOKEN_DECIMALS } from "./config.js";
import { CANCEL_RFQ_QUOTE, CANCEL_RFQ_REQUEST, CREATE_RFQ_QUOTE, CREATE_RFQ_REQUEST, GET_RFQ_BEST_QUOTE, GET_RFQ_QUOTER_QUOTES, GET_RFQ_REQUESTER_QUOTES, GET_RFQ_REQUESTS, RFQ_CONFIG, RFQ_QUOTE_APPROVE, RFQ_REQUESTS_ACCEPT, } from "./endpoints.js";
import { L1_AUTH_UNAVAILABLE_ERROR, L2_AUTH_NOT_AVAILABLE } from "./errors.js";
import { createL2Headers } from "./headers/index.js";
import { DELETE, GET, POST } from "./http-helpers/index.js";
import { ROUNDING_CONFIG } from "./order-builder/helpers.js";
import { RfqMatchType, Side } from "./types.js";
import { roundDown, roundNormal } from "./utilities.js";
// RFQ list params need to be repeated e.g. quoteIds=...&quoteIds=...
const buildRepeatedQuery = (params) => {
    const sp = new URLSearchParams();
    if (!params)
        return "";
    Object.entries(params).forEach(([key, value]) => {
        if (value === undefined || value === null) {
            return;
        }
        if (Array.isArray(value)) {
            for (const v of value)
                sp.append(key, String(v));
        }
        else {
            sp.append(key, String(value));
        }
    });
    return sp.toString();
};
/**
 * RfqClient provides RFQ (Request for Quote) functionality on top of a CLOB client.
 */
export class RfqClient {
    deps;
    constructor(deps) {
        this.deps = deps;
    }
    /**
     * Creates and posts an RFQ request from user order parameters.
     * Converts the order into the proper RFQ request format with asset amounts and submits it.
     */
    async createRfqRequest(userOrder, options) {
        this.ensureL2Auth();
        const { tokenID, price, side, size } = userOrder;
        const tickSize = await this.deps.resolveTickSize(tokenID, options?.tickSize);
        const roundConfig = ROUNDING_CONFIG[tickSize];
        const roundedPrice = roundNormal(price, roundConfig.price).toFixed(roundConfig.price);
        const roundedSize = roundDown(size, roundConfig.size).toFixed(roundConfig.size);
        // Calculate amounts based on side
        const sizeNum = Number.parseFloat(roundedSize);
        const priceNum = Number.parseFloat(roundedPrice);
        let amountIn;
        let amountOut;
        let assetIn;
        let assetOut;
        if (side === Side.BUY) {
            // Buying: pay USDC (asset 0), receive tokens (tokenID)
            amountIn = parseUnits(roundedSize, COLLATERAL_TOKEN_DECIMALS).toString();
            amountOut = parseUnits((sizeNum * priceNum).toFixed(roundConfig.amount), COLLATERAL_TOKEN_DECIMALS).toString();
            assetIn = tokenID;
            assetOut = "0"; // USDC
        }
        else {
            // Selling: pay tokens (tokenID), receive USDC (asset 0)
            amountIn = parseUnits((sizeNum * priceNum).toFixed(roundConfig.amount), COLLATERAL_TOKEN_DECIMALS).toString();
            amountOut = parseUnits(roundedSize, COLLATERAL_TOKEN_DECIMALS).toString();
            assetIn = "0"; // USDC
            assetOut = tokenID;
        }
        const payload = {
            assetIn,
            assetOut,
            amountIn,
            amountOut,
            userType: this.deps.userType,
        };
        const endpoint = CREATE_RFQ_REQUEST;
        const l2HeaderArgs = {
            method: POST,
            requestPath: endpoint,
            body: JSON.stringify(payload),
        };
        const headers = await createL2Headers(this.deps.signer, this.deps.creds, l2HeaderArgs, this.deps.useServerTime ? await this.deps.getServerTime() : undefined);
        return this.deps.post(`${this.deps.host}${endpoint}`, { headers, data: payload });
    }
    /**
     * Cancels an existing RFQ request
     */
    async cancelRfqRequest(request) {
        this.ensureL2Auth();
        const endpoint = CANCEL_RFQ_REQUEST;
        const l2HeaderArgs = {
            method: DELETE,
            requestPath: endpoint,
            body: JSON.stringify(request),
        };
        const headers = await createL2Headers(this.deps.signer, this.deps.creds, l2HeaderArgs, this.deps.useServerTime ? await this.deps.getServerTime() : undefined);
        return this.deps.del(`${this.deps.host}${endpoint}`, { headers, data: request });
    }
    /**
     * Gets RFQ requests with optional filtering parameters
     */
    async getRfqRequests(params) {
        this.ensureL2Auth();
        const endpoint = GET_RFQ_REQUESTS;
        const l2HeaderArgs = {
            method: GET,
            requestPath: endpoint,
        };
        const headers = await createL2Headers(this.deps.signer, this.deps.creds, l2HeaderArgs, this.deps.useServerTime ? await this.deps.getServerTime() : undefined);
        const query = buildRepeatedQuery(params);
        const url = query
            ? `${this.deps.host}${endpoint}?${query}`
            : `${this.deps.host}${endpoint}`;
        return this.deps.get(url, { headers });
    }
    /**
     * Creates a quote in response to an RFQ request.
     * Converts user-friendly parameters into the proper RFQ quote format with asset amounts and submits it.
     */
    async createRfqQuote(userQuote, options) {
        this.ensureL2Auth();
        const { requestId, tokenID, price, side, size } = userQuote;
        const tickSize = await this.deps.resolveTickSize(tokenID, options?.tickSize);
        const roundConfig = ROUNDING_CONFIG[tickSize];
        const roundedPrice = roundNormal(price, roundConfig.price).toFixed(roundConfig.price);
        const roundedSize = roundDown(size, roundConfig.size).toFixed(roundConfig.size);
        // Calculate amounts based on side
        const sizeNum = Number.parseFloat(roundedSize);
        const priceNum = Number.parseFloat(roundedPrice);
        let amountIn;
        let amountOut;
        let assetIn;
        let assetOut;
        if (side === Side.SELL) {
            // Quoter selling tokens: receive USDC (asset 0), give tokens (tokenID)
            amountIn = parseUnits((sizeNum * priceNum).toFixed(roundConfig.amount), COLLATERAL_TOKEN_DECIMALS).toString();
            amountOut = parseUnits(roundedSize, COLLATERAL_TOKEN_DECIMALS).toString();
            assetIn = "0"; // USDC
            assetOut = tokenID;
        }
        else {
            // Quoter buying tokens: receive tokens (tokenID), give USDC (asset 0)
            amountIn = parseUnits(roundedSize, COLLATERAL_TOKEN_DECIMALS).toString();
            amountOut = parseUnits((sizeNum * priceNum).toFixed(roundConfig.amount), COLLATERAL_TOKEN_DECIMALS).toString();
            assetIn = tokenID;
            assetOut = "0"; // USDC
        }
        const payload = {
            requestId,
            assetIn,
            assetOut,
            amountIn,
            amountOut,
        };
        const endpoint = CREATE_RFQ_QUOTE;
        // Auto-fill userType from client's signatureType
        const quoteWithUserType = {
            ...payload,
            userType: this.deps.userType,
        };
        const l2HeaderArgs = {
            method: POST,
            requestPath: endpoint,
            body: JSON.stringify(quoteWithUserType),
        };
        const headers = await createL2Headers(this.deps.signer, this.deps.creds, l2HeaderArgs, this.deps.useServerTime ? await this.deps.getServerTime() : undefined);
        return this.deps.post(`${this.deps.host}${endpoint}`, { headers, data: quoteWithUserType });
    }
    /**
     * Gets quotes on requests created by the authenticated user (requester view).
     * Returns quotes that others have made on your requests.
     */
    async getRfqRequesterQuotes(params) {
        this.ensureL2Auth();
        const endpoint = GET_RFQ_REQUESTER_QUOTES;
        const l2HeaderArgs = {
            method: GET,
            requestPath: endpoint,
        };
        const headers = await createL2Headers(this.deps.signer, this.deps.creds, l2HeaderArgs, this.deps.useServerTime ? await this.deps.getServerTime() : undefined);
        const query = buildRepeatedQuery(params);
        const url = query
            ? `${this.deps.host}${endpoint}?${query}`
            : `${this.deps.host}${endpoint}`;
        return this.deps.get(url, { headers });
    }
    /**
     * Gets quotes created by the authenticated user (quoter view).
     * Returns quotes that you have made on others' requests.
     */
    async getRfqQuoterQuotes(params) {
        this.ensureL2Auth();
        const endpoint = GET_RFQ_QUOTER_QUOTES;
        const l2HeaderArgs = {
            method: GET,
            requestPath: endpoint,
        };
        const headers = await createL2Headers(this.deps.signer, this.deps.creds, l2HeaderArgs, this.deps.useServerTime ? await this.deps.getServerTime() : undefined);
        const query = buildRepeatedQuery(params);
        const url = query
            ? `${this.deps.host}${endpoint}?${query}`
            : `${this.deps.host}${endpoint}`;
        return this.deps.get(url, { headers });
    }
    /**
     * Gets the best quote for a given request ID
     */
    async getRfqBestQuote(params) {
        this.ensureL2Auth();
        const endpoint = GET_RFQ_BEST_QUOTE;
        const l2HeaderArgs = {
            method: GET,
            requestPath: endpoint,
        };
        const headers = await createL2Headers(this.deps.signer, this.deps.creds, l2HeaderArgs, this.deps.useServerTime ? await this.deps.getServerTime() : undefined);
        return this.deps.get(`${this.deps.host}${endpoint}`, { headers, params });
    }
    /**
     * Cancels an existing quote
     */
    async cancelRfqQuote(quote) {
        this.ensureL2Auth();
        const endpoint = CANCEL_RFQ_QUOTE;
        const l2HeaderArgs = {
            method: DELETE,
            requestPath: endpoint,
            body: JSON.stringify(quote),
        };
        const headers = await createL2Headers(this.deps.signer, this.deps.creds, l2HeaderArgs, this.deps.useServerTime ? await this.deps.getServerTime() : undefined);
        return this.deps.del(`${this.deps.host}${endpoint}`, { headers, data: quote });
    }
    /**
     * Gets the RFQ configuration from the server
     */
    async rfqConfig() {
        this.ensureL2Auth();
        const endpoint = RFQ_CONFIG;
        const l2HeaderArgs = {
            method: GET,
            requestPath: endpoint,
        };
        const headers = await createL2Headers(this.deps.signer, this.deps.creds, l2HeaderArgs, this.deps.useServerTime ? await this.deps.getServerTime() : undefined);
        return this.deps.get(`${this.deps.host}${endpoint}`, { headers });
    }
    /**
     * Accepts a quote and creates an order (taker side)
     * This fetches the request details, creates an order, and submits the acceptance
     */
    async acceptRfqQuote(payload) {
        this.ensureL2Auth();
        const rfqQuotes = await this.getRfqRequesterQuotes({
            quoteIds: [payload.quoteId],
        });
        // Check for HTTP errors first (network failures, auth errors, server errors)
        if ("error" in rfqQuotes) {
            const errorMsg = typeof rfqQuotes.error === "string"
                ? rfqQuotes.error
                : JSON.stringify(rfqQuotes.error);
            throw new Error(`Error fetching RFQ quote: ${errorMsg}`);
        }
        if (!rfqQuotes?.data || rfqQuotes.data.length === 0) {
            throw new Error("RFQ quote not found");
        }
        const rfqQuote = rfqQuotes.data[0];
        // Create an order, matching the quote
        const orderPayload = this.getRequestOrderCreationPayload(rfqQuote);
        const orderSide = orderPayload.side;
        const size = orderPayload.size;
        const token = orderPayload.token;
        const order = await this.deps.createOrder({
            tokenID: token,
            price: orderPayload.price,
            size: Number.parseFloat(size),
            side: orderSide,
            expiration: payload.expiration,
        });
        if (!order) {
            throw new Error("Error creating order");
        }
        const acceptPayload = {
            requestId: payload.requestId,
            quoteId: payload.quoteId,
            owner: this.deps.creds.key,
            ...order,
            expiration: Number.parseInt(order.expiration, 10),
            side: orderSide,
            salt: Number.parseInt(order.salt.toString(), 10),
        };
        const endpoint = RFQ_REQUESTS_ACCEPT;
        const l2HeaderArgs = {
            method: POST,
            requestPath: endpoint,
            body: JSON.stringify(acceptPayload),
        };
        const headers = await createL2Headers(this.deps.signer, this.deps.creds, l2HeaderArgs, this.deps.useServerTime ? await this.deps.getServerTime() : undefined);
        return this.deps.post(`${this.deps.host}${endpoint}`, { headers, data: acceptPayload });
    }
    /**
     * Approves a quote and creates an order (maker side)
     * This fetches the quote details, creates an order, and submits the approval
     */
    async approveRfqOrder(payload) {
        this.ensureL2Auth();
        const rfqQuotes = await this.getRfqQuoterQuotes({
            quoteIds: [payload.quoteId],
        });
        // Check for HTTP errors first (network failures, auth errors, server errors)
        if ("error" in rfqQuotes) {
            const errorMsg = typeof rfqQuotes.error === "string"
                ? rfqQuotes.error
                : JSON.stringify(rfqQuotes.error);
            throw new Error(`Error fetching RFQ quote: ${errorMsg}`);
        }
        if (!rfqQuotes?.data || rfqQuotes.data.length === 0) {
            throw new Error("RFQ quote not found");
        }
        const rfqQuote = rfqQuotes.data[0];
        // Create an order based on the quote details
        const side = rfqQuote.side === "BUY" ? Side.BUY : Side.SELL;
        const size = rfqQuote.side === "BUY" ? rfqQuote.sizeIn : rfqQuote.sizeOut;
        const order = await this.deps.createOrder({
            tokenID: rfqQuote.token,
            price: rfqQuote.price,
            size: Number.parseFloat(size),
            side: side,
            expiration: payload.expiration,
        });
        if (!order) {
            throw new Error("Error creating order");
        }
        const approvePayload = {
            requestId: payload.requestId,
            quoteId: payload.quoteId,
            owner: this.deps.creds.key,
            ...order,
            expiration: Number.parseInt(order.expiration, 10),
            side: side,
            salt: Number.parseInt(order.salt.toString(), 10),
        };
        const endpoint = RFQ_QUOTE_APPROVE;
        const l2HeaderArgs = {
            method: POST,
            requestPath: endpoint,
            body: JSON.stringify(approvePayload),
        };
        const headers = await createL2Headers(this.deps.signer, this.deps.creds, l2HeaderArgs, this.deps.useServerTime ? await this.deps.getServerTime() : undefined);
        return this.deps.post(`${this.deps.host}${endpoint}`, { headers, data: approvePayload });
    }
    /**
     * Ensures L2 authentication is available for RFQ endpoints.
     */
    ensureL2Auth() {
        if (this.deps.signer === undefined) {
            throw L1_AUTH_UNAVAILABLE_ERROR;
        }
        if (this.deps.creds === undefined) {
            throw L2_AUTH_NOT_AVAILABLE;
        }
    }
    getRequestOrderCreationPayload(quote) {
        const quoteSide = quote.side;
        const matchType = quote.matchType;
        let side;
        let token;
        let size;
        let price;
        switch (matchType) {
            case RfqMatchType.COMPLEMENTARY:
                // For BUY <> SELL and SELL <> BUY
                // the order side is opposite the quote side
                side = quoteSide === "BUY" ? Side.SELL : Side.BUY;
                token = quote.token;
                size = side === Side.BUY ? quote.sizeOut : quote.sizeIn;
                price = quote.price;
                return {
                    token,
                    side,
                    size,
                    price,
                };
            case RfqMatchType.MINT:
            case RfqMatchType.MERGE:
                // BUY<> BUY, SELL <> SELL
                // the order side is the same as the quote side
                side = quoteSide === "BUY" ? Side.BUY : Side.SELL;
                token = quote.complement;
                size = side === Side.BUY ? quote.sizeIn : quote.sizeOut;
                // For a MINT or a MERGE, the requester price is the inverse of the quote price
                // 95c Quote to BUY NO, implies that the Requester is buying YES at 5c
                // 45c Quote to SELL NO, implies the Requester is selling YES at 55c
                price = 1 - quote.price;
                return {
                    token,
                    side,
                    size,
                    price,
                };
            default:
                throw new Error("invalid match type");
        }
    }
}
