// @polymarket/clob-client/dist/signing/eip712.js
import { getSignerAddress, signTypedDataWithSigner } from "../signer.js";
import { MSG_TO_SIGN } from "./constants.js";
/**
 * Builds the canonical Polymarket CLOB EIP712 signature
 * @param signer
 * @param ts
 * @returns string
 */
export const buildClobEip712Signature = async (signer, chainId, timestamp, nonce) => {
    const address = await getSignerAddress(signer);
    const ts = `${timestamp}`;
    const domain = {
        name: "ClobAuthDomain",
        version: "1",
        chainId: chainId,
    };
    const types = {
        ClobAuth: [
            { name: "address", type: "address" },
            { name: "timestamp", type: "string" },
            { name: "nonce", type: "uint256" },
            { name: "message", type: "string" },
        ],
    };
    const value = {
        address,
        timestamp: ts,
        nonce,
        message: MSG_TO_SIGN,
    };
    const sig = await signTypedDataWithSigner({
        signer,
        domain,
        types,
        value,
        primaryType: "ClobAuth",
    });
    return sig;
};
