// @polymarket/clob-client/dist/order-utils/exchange.order.builder.js
import { hashTypedData } from "viem";
import { getSignerAddress, signTypedDataWithSigner } from "../signer.js";
import { EIP712_DOMAIN, ORDER_STRUCTURE, PROTOCOL_NAME, PROTOCOL_VERSION, } from "./exchange.order.const.js";
import { SignatureType } from "./model/signature-types.model.js";
import { generateOrderSalt } from "./utils.js";
export class ExchangeOrderBuilder {
    contractAddress;
    chainId;
    signer;
    generateSalt;
    constructor(contractAddress, chainId, signer, generateSalt = generateOrderSalt) {
        this.contractAddress = contractAddress;
        this.chainId = chainId;
        this.signer = signer;
        this.generateSalt = generateSalt;
    }
    /**
     * Build an order object including the signature.
     */
    async buildSignedOrder(orderData) {
        const order = await this.buildOrder(orderData);
        const orderTypedData = this.buildOrderTypedData(order);
        const orderSignature = await this.buildOrderSignature(orderTypedData);
        return {
            ...order,
            signature: orderSignature,
        };
    }
    /**
     * Creates an Order object from order data.
     */
    async buildOrder({ maker, taker, tokenId, makerAmount, takerAmount, side, feeRateBps, nonce, signer, expiration, signatureType, }) {
        if (typeof signer === "undefined" || !signer) {
            signer = maker;
        }
        const signerAddress = await getSignerAddress(this.signer);
        if (signer !== signerAddress) {
            throw new Error("signer does not match");
        }
        if (typeof expiration === "undefined" || !expiration) {
            expiration = "0";
        }
        if (typeof signatureType === "undefined" || !signatureType) {
            // Default to EOA 712 sig type
            signatureType = SignatureType.EOA;
        }
        return {
            salt: this.generateSalt(),
            maker,
            signer,
            taker,
            tokenId,
            makerAmount,
            takerAmount,
            expiration,
            nonce,
            feeRateBps,
            side,
            signatureType,
        };
    }
    /**
     * Parses an Order object to EIP712 typed data.
     */
    buildOrderTypedData(order) {
        return {
            primaryType: "Order",
            types: {
                EIP712Domain: EIP712_DOMAIN,
                Order: ORDER_STRUCTURE,
            },
            domain: {
                name: PROTOCOL_NAME,
                version: PROTOCOL_VERSION,
                chainId: this.chainId,
                verifyingContract: this.contractAddress,
            },
            message: {
                salt: order.salt,
                maker: order.maker,
                signer: order.signer,
                taker: order.taker,
                tokenId: order.tokenId,
                makerAmount: order.makerAmount,
                takerAmount: order.takerAmount,
                expiration: order.expiration,
                nonce: order.nonce,
                feeRateBps: order.feeRateBps,
                side: order.side,
                signatureType: order.signatureType,
            },
        };
    }
    /**
     * Generates order signature from EIP712 typed data.
     */
    async buildOrderSignature(typedData) {
        const { EIP712Domain: _, ...orderTypes } = typedData.types;
        return signTypedDataWithSigner({
            signer: this.signer,
            domain: typedData.domain,
            types: orderTypes,
            value: typedData.message,
            primaryType: typedData.primaryType,
        });
    }
    /**
     * Generates the hash of the order from EIP712 typed data.
     */
    buildOrderHash(orderTypedData) {
        const { EIP712Domain: _, ...orderTypes } = orderTypedData.types;
        return hashTypedData({
            domain: orderTypedData.domain,
            types: orderTypes,
            primaryType: orderTypedData.primaryType,
            message: orderTypedData.message,
        });
    }
}
